initial draft of user login endpoint

Sullivan-Patrick opened the pull request.

Sullivan-Patrick committed:
initial draft of user login endpoint, incomplete + no tests yet

Sullivan-Patrick committed:
Drop files from .gitignore

Sullivan-Patrick committed:
Drop files from .gitignore

Sullivan-Patrick committed:
Drop files from .gitignore

astherath commented on docs/users.py:
```diff
@@ -28,3 +28,10 @@
 Find user by identifier
 """
+
+validate_user_desc = """
+Validates user against database using 
+"""
+validate_user_summ = """
+Validates user
+"""
```
Docs names need to be changed to fit new naming convention for the endpoint

astherath commented on models/users.py:
```diff
@@ -84,3 +84,18 @@ class UserInfoQueryResponse(User):
     """
+
+
+class ValidateUserForm(BaseModel):
+    """
+    Contains username and password to validate against database
+    """
+    identifier: UserIdentifier
+    password: str
+
+
+class ValidateLoginResponse(BaseModel):
+    """
+    Response for a user login attempt
+    """
+    jwt: str
```
Classes are perfect, but naming should reflect new endpoint. Maybe `UserLoginForm` and `UserLoginResponse`?

astherath commented on routes/users.py:
```diff
@@ -48,3 +48,17 @@ async def delete_user(identifier: models.UserIdentifier):
     return models.UserInfoQueryResponse(**user_data.dict())
+
+
+# this implementation doesn't use models.ValidateUserResponse. Should I implement this?
```
Yep. Whatever data the method below returns will be cast into a `ValidateLoginResponse`, so if the data you're returning _can't_ be cast, you'll get an error. 

The easiest way to do this is to make sure that the `utils` method that you're calling in the function below returns a `ValidateLoginResponse` model.

astherath commented on routes/users.py:
```diff
@@ -48,3 +48,17 @@ async def delete_user(identifier: models.UserIdentifier):
+
+
+# this implementation doesn't use models.ValidateUserResponse. Should I implement this?
+# also, how can I define a 404 response if it user isn't found
```
Good question. We can't actually define ahead of time what exceptions can be thrown, but we can just throw them as normal and they will be interpreted by the FastAPI code as `HTTPException`s and behave as expected in the frontend.

Take a look in the `models.exceptions` module for an example of an exception if you need a custom one. 
If one of the exceptions in that module fits your needs, then you can just `raise` it and it'll go all the way to the client and work as you expect. 

e.g.:
```python
if not user_document:
    raise exceptions.UserNotFoundException
```

astherath commented on routes/users.py:
```diff
@@ -48,3 +48,17 @@ async def delete_user(identifier: models.UserIdentifier):
+
+# this implementation doesn't use models.ValidateUserResponse. Should I implement this?
+# also, how can I define a 404 response if it user isn't found
+@router.post("/users/validate",  # todo: finish this endpoint
+             response_model=models.ValidateLoginResponse,
+             description=docs.validate_user_desc,
+             summary=docs.validate_user_summ,
+             tags=["Users"],
+             status_code=200)
+async def validate_user(validate: models.ValidateUserForm):  # todo: figure out how this works
+    return await utils.validate_user_info(validate)
```
This code looks perfect. 

But here's a broader design question to worry about: is `validate` the best way to call this flow? 
In a micro sense, we are just validating a user by their password, but that doesn't really mean much. Does `validate_user` validate the user's data? Does it validate that the user _exists_? 

If what we're doing is validating the user's password, then we should use the common name for that: `user_login`.

This makes it instantly obvious to anyone who needs to read the code, as well as being a universal workflow that _anyone_ can understand.

While it might seems annoying and unnecessary, small changes like these keep the code clean and healthy.

Also, no one wants to be the one to have to explain to frontend why their login endpoint isn't called login hahaha.

There are other comments mimicking this one in the modules that also need name changes, use them as reminders for what code needs to be fixed.

astherath commented on util/users.py:
```diff
@@ -55,3 +55,21 @@ async def delete_user(identifier: user_models.UserIdentifier) -> None:
         raise exceptions.UserNotFoundException(detail=detail)
+
+
+async def validate_user_info(
```
should be called something along the lines of `try_to_login_user` or `attempt_user_login` or something that tells what the function does, even if it is more verbose.

astherath commented on util/users.py:
```diff
@@ -55,3 +55,21 @@ async def delete_user(identifier: user_models.UserIdentifier) -> None:
+
+
+async def validate_user_info(
+        validate: user_models.ValidateUserForm) -> str:
```
Consider changing the return type (and type hints) to be `user_models.ValidateUserResponse` (post `login`-refactor) for convenience of the caller.

astherath replied:
`validate` is a non-descriptive variable name. Something like `login_form` or `login_data` would be much better.

astherath commented on util/users.py:
```diff
@@ -55,3 +55,21 @@ async def delete_user(identifier: user_models.UserIdentifier) -> None:
+    Validates user based off identifier and password
+    #todo: call get_user_info_by_identifier w identifier, validate pass vs user object. return 200
+    """
+    try:
+        user_info = await get_user_info_by_identifier(validate.identifier)
+    except exceptions.UserNotFoundException:
+        return 'not found'
```
If the goal of this try/except block is to account for non-existent users, then it is actually redundant.

If you see the code in `get_user_info_by_identifier`, it will raise it's own `exception` (the same one we're catching here). So this is effectively catching it to re-raise it.

Consider instead just `user_info = await get_user_info_by_identifier(validate.identifier)`, which will raise a 404 by itself anyway.

NOTE: this isn't totally clear in the `get_user_info_by_identifier` docstring, but it should be. Bonus points if you add a line or two to that method's documentation that details the fact that it will raise a 404 `UserNotFoundException` if no user is found (boy scout rule).

astherath commented on util/users.py:
```diff
@@ -55,3 +55,21 @@ async def delete_user(identifier: user_models.UserIdentifier) -> None:
+        return 'not found'
+
+    # if user_info.hashed_password == hash_pass(validate.password):
+    if validate.password == 'password123':
+        return 'token here'
+    else:
+        return 'not found'
```
This is obviously placeholder code, but it will not be easy or clean to refactor.

A better approach would be offloading the password comparison to a function so _when_ we need to actually implement it, this code will be the exact same.

e.g.
```python
password_matches = check_user_password_matches(login_form, user_info) # should return bool
if password_matches:
    return get_auth_token_from_user_data(user_data)
else:
    raise exceptions.InvalidPasswordException # this doesn't exist yet but it should throw a 422 or 403
```

This way, when we want to change anything about password comparison we can easily just modify the `check_user_password_matches` function. This is the core of making clean code: how easy it is to refactor for improvement.

Sullivan-Patrick replied:
Isn't it bad practice to return a different response when the password is wrong versus when the user doesn't exist?

astherath replied:
Agreed, _returning_ something different is bad, but we're actually not returning something different if the password is invalid, we're raising an exception. 

Only the "successful" control flow should return an object, all others should just raise exceptions (they go all the way to the top of the stack to the client).

astherath reviewed (changes requested):
Great initial push! 

Changes requested mostly revolve around renaming `validation_` to `login_` module-wide for clarity.

Also left some refactoring comments on the main util handler method to make it easier to implement the incoming auth token code.

As a final note, there's some lint errors in the code:
```
routes/users.py:53:0: C0301: Line too long (87/80) (line-too-long)
routes/users.py:61:0: C0301: Line too long (94/80) (line-too-long)
************* Module util.users
util/users.py:64:0: C0301: Line too long (98/80) (line-too-long)
util/users.py:72:4: R1705: Unnecessary "else" after "return" (no-else-return)
util/users.py:67:8: W0612: Unused variable 'user_info' (unused-variable)
```
Try to run the linter before every push just so the CI can give you the nice green checkmark instead of failing, for more info on the lint comments/errors check the [pylint](http://pylint.pycqa.org/en/latest/technical_reference/features.html#basic-checker-messages) docs

p.s: dont forget to write some tests!!

Sullivan-Patrick committed:
finished implementing base code, started testing

Sullivan-Patrick committed:
finished implementing base code, started testing

Sullivan-Patrick committed:
Merge branch 'master' of https://github.com/SparkDevTeams/underline-backend into feature-user-login-endpoint

Sullivan-Patrick committed:
Spruced up some functions to include for password field

Sullivan-Patrick committed:
worked more on tests

astherath committed:
added missing code for failing test

Sullivan-Patrick committed:
Tests now work, minor formatting fixes

Sullivan-Patrick committed:
Some formatting, changed lint script to ignore venv

Sullivan-Patrick committed:
formatting. Won't pass pylint because user is still unused in dummy function

astherath commented on docs/users.py:
```diff
@@ -28,3 +28,10 @@
 """
+
+login_user_desc = """
+Validates user against database using 
+"""
+login_user_summ = """
+Validates user
+"""
```
Docs should be updated to reflect the name change to `login`.

Sullivan-Patrick replied:
Since the login process _is_ validation, would this suffice?

login_user_desc = """
Validates and logs in a user against database
"""
login_user_summ = """
Validates and logs in a user
"""

astherath replied:
Since these docs are going to the frontend team anyway we might as well speak their language; at least for the `summ`, "User login" would be more than good.

astherath commented on routes/users.py:
```diff
@@ -48,3 +48,14 @@ async def delete_user(identifier: models.UserIdentifier):
+             status_code=200)
+async def login_user(validate: models.UserLoginForm):
+    print('routes')
+    return await utils.attempt_user_login(validate)
```
verbs as variable names is a code smell in most cases, including this one.

`validate` as the name of the incoming data is pretty nondescript. Since the scope of `validate` is pretty tiny we don't need to go overboard, but a more conventional name like `login_form` or even just `form_data` would be better.

astherath commented on routes/users.py:
```diff
@@ -48,3 +48,14 @@ async def delete_user(identifier: models.UserIdentifier):
+             tags=["Users"],
+             status_code=200)
+async def login_user(validate: models.UserLoginForm):
+    print('routes')
```
dead code/old debug statement.

Sullivan-Patrick replied:
just the print statement is dead code, right?

astherath replied:
yep

astherath commented on tests/test_attempt_user_login.py:
```diff
@@ -0,0 +1,93 @@
+# pylint: disable=logging-fstring-interpolation
+#       - honestly just annoying to use lazy(%) interpolation.
+"""
+Endpoint testing for deleting feedback off of an event.
```
Wrong docstring (copy paste responsibly !)

Sullivan-Patrick replied:
me copy pasting 

https://media3.giphy.com/media/qY8qEkjdUIsKIPCMCl/giphy.gif?cid=ecf05e47t56rvf3uxt11sdyfmcjpg88z3xr80yzewwwhfx9f&rid=giphy.gif

astherath commented on tests/test_attempt_user_login.py:
```diff
@@ -0,0 +1,93 @@
+"""
+
+
+def get_user_login_endpoint_url() -> str:  # pylint: disable=invalid-name
```
unnecessary `pylint: disable`.

astherath commented on tests/test_attempt_user_login.py:
```diff
@@ -0,0 +1,93 @@
+    Helper function that checks if status code is valid
+    """
+    try:
+        assert response.status_code == 200
```
this is a good start, but we should assert some more things to make sure the data we're getting back is good.

Maybe asserting that there's a token in the response? Later on this will be extended to make sure the token we receive is valid as well.

Sullivan-Patrick replied:
I'm having trouble figuring out how to assert there is a token in the response. The response is in the form of a `UserLoginResponse` object, yeah? So that means it will be json that includes strictly the jwt? But I'm not sure how to validate if what it returns _is_ a jwt, or do I just need to check the jwt field was actually filled out?

astherath replied:
We don't have the code that has the ability to check if a `jwt` is valid just yet, so just checking for existence is good enough at this point.

When that code gets into upstream we'll refactor this to check for `jwt` validity as well.

astherath commented on tests/test_attempt_user_login.py:
```diff
@@ -0,0 +1,93 @@
+        """
+        request_url = get_user_login_endpoint_url()
+        json_payload = get_user_login_json_data(registered_user)
+        identifier = json_payload.get("identifier")
+        false_pass = 'aaaaaaaaaa'
+        assert false_pass != json_payload.get("password")
+        incorrect_pass_payload = {"identifier": identifier,
+                                  "password": false_pass}
```
Kinda unnecessary amount of work; just changing the `json_payload["password"]` to a random string would be good enough.

Sullivan-Patrick replied:
For some reason I thought dicts were immutable. Nice.

astherath commented on util/users.py:
```diff
@@ -55,3 +56,34 @@ async def delete_user(identifier: user_models.UserIdentifier) -> None:
+    raise exceptions.InvalidPasswordException
+
+
+# fixme: change this to use hashed passwords once available
+async def check_user_password_matches(login_form: user_models.UserLoginForm,
+                                      user: user_models.User) -> bool:
+    return login_form.password == user.password
```
Code to fix this is in upstream.

Also, since you offloaded the password comparison to this function, the refactor will be extremely easy, which is a great sign for the code quality.

Sullivan-Patrick replied:
-I vaguely remember us discussing this, but I believe my code is up to date with upstream (I see last change was 10 days ago, Christian's push). Regardless, how can I pull master into my codebase without compromising what I've written? Was it as simple as doing a `git pull origin master`, or was it something else?

-`user.password` is going to be hashed, right? So the code would just be return user.check_password(login_form.password)? If that assumption is correct, that seems _almost_ overly complex, since user already has a `check_password` function. What do you think?

astherath replied:
Yeah your code is up to date with master (we ran `git pull origin master` last meeting), so this can implement the actual function.

Assumption is definitely correct, we never touch the hashed passwords directly so we would just be calling `user.check_password(login_form.password)` and returning that.

While plugging that call into the main `attempt_user_login` function is totally valid, it makes us call a method for a class that we don't have to interact with directly in the function, essentially making a mixed level of abstraction in the code. 

What if, for example, once we change the `User` class to also allow admin type users, we have a more complex password checking system that also checks for roles or something like that. 

By abstracting the actual call to the `User.check_password` method in it's own function, we would be able to rewrite that single function and never have to change the actual login function code.

Sometimes writing refactorable code seems a little extra and almost redundant, but adding clarity and flexibility is always worth the extra code. Writing more code is bad if it's just more code, but it almost never is.

Sullivan-Patrick replied:
True, your example makes a lot of sense. I know it was a clarity/refactorability tradeoff, but considering the admin stuff is coming soon I think it makes a lot of sense to add this extra layer here.

astherath reviewed (changes requested):
Good stuff, small refactorings suggested as well as a missing password comparison. 

Also I think there was a lint fix missing for the CI/CD to pass, so just check that out.

Sullivan-Patrick committed:
Merge branch 'master' of https://github.com/SparkDevTeams/underline-backend into feature-user-login-endpoint

Sullivan-Patrick committed:
Made (hopefully) final fixes. Added a dummy check_jwt_valid function

astherath committed:
fixed user reg bug with password not being set.

The bug was actually just a missing implementation of the  method on incoming user registration.

Once added, the  fixture also had to get updated so it would return the pre-hash password to pass tests

astherath committed:
refactored util comparison method to pass test.

the response - registered_user comparison method in  relied on
comparing all fields, including , which is now hashed/unhashed as per the
new fixture. This refactor allows for a flexible set of blacklisted fields to not compare.

astherath commented:
LGTM!
